main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,381 @@
+        case Pattern.Int32(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Int64(lit, _) => acc ++ ", " + lit.toString
+        case Pattern.Float32(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Float64(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Char(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.BigInt(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Wild(_, _) => acc ++ ", _"
+        case Pattern.Var(_, _, _) => acc ++ ", _"
+        case Pattern.Tuple(pats, _ ,_ ) => acc ++ ", (" ++ formatRule(pats) ++ ")"
+        case Pattern.Tag(sym, tag, pt, _, _) => acc ++ ", " ++ sym.name ++ "/" ++ tag ++ formatRule(List(pt))
+      }
+      rule.foldRight("")(formatPattern)
+    }
+
+    /**
+      * Given a list of patterns, computes a pattern vector of size n such
+      * that p doesn't match any rule in rules
+      *
+      * @param  root The AST root of the expression
+      * @param  exp The expression the patterns match against
+      * @param rules The rules to check for exhaustion
+      * @param n The size of resulting pattern vector
+      * @return If no such pattern exists, returns Right(the input), else returns Left(a matching pattern)
+      */
+    def FindNonMatchingPat(root: TypedAst.Root, exp: TypedAst.Expression, rules: List[List[TypedAst.Pattern]], n: Integer): Either[(EnumSym, List[Pattern]), List[List[Pattern]]] = {
+      if (n==0) {
+        if (rules.isEmpty) {
+          return Left((new EnumSym(List.empty[String], "", null), List.empty[Pattern]))
+        }
+        return Right(rules)
+      }
+
+      val sigma = rootCtors(rules)
+      if (completeSig(root, exp, sigma)) {
+        /* If the constructors are complete, then we check that the arguments to the constructors and the remaining
+         * patterns are complete
+         *
+         * e.g. If we have
+         * enum Option {
+         *    case Some a,
+         *    case Nothing
+         *  }
+         * case Some True =>
+         * case Some False =>
+         * case Nothing =>
+         *
+         * {Some, Nothing} is a complete signature, but the just case is exhaustive only if the {True, False} case is
+         *
+         * exhaustive. So we create a "Specialized" matrix for Just with {True, False} as rows and check that.
+         */
+        val check_all: List[Either[(EnumSym, List[Pattern]), List[List[Pattern]]]] = sigma.map(c => FindNonMatchingPat(root, exp, Specialize(root, c, rules),  countCtorArgs(c) + n - 1))
+        val base: Either[(EnumSym, List[Pattern]), List[List[Pattern]]] = Right(List.empty[List[Pattern]])
+        val rules_valid: Either[(EnumSym, List[Pattern]), List[List[Pattern]]] = check_all.foldRight(base)(mergeEither)
+        rules_valid match {
+          case Right(a) => Right(a)
+          case Left((c, pats)) => Left((c, List(TypedAst.Pattern.Tag(c, c.name, TypedAst.Pattern.Tuple(pats, exp.tpe, exp.loc), exp.tpe, exp.loc))))
```

Comment:
Is there any reason to *rebuild* the AST? I.e. you are constructing a new Typed.Pattern.Tag object.